### PR TITLE
Fix flaky text tests

### LINKED ISSUE #7911: mathtext/mathfont intermittent failures
I'm opening this as an issue separate from #6586 to better document the investigation into the root cause. This affects both the `v2.0.x` and `master` branches.

We can make the failures a bit more reproducible by repeating the tests a bit. What I did was the following:
1. Choose one of the mathtext/mathfont tests (e.g. `mathfont_*_08`)
2. Copy that test image 100 times for each font set 
  ```bash
for i in `seq -w 0 99`; do
    for fs in cm dejavusans dejavuserif stix stixsans; do
        cp test_mathtext/mathfont_${fs}_08.png test_fail/mathfont_${fs}_${i}.png
    done
done
```
3. Modify `test_mathtext.py` to drop the other tests and run only `08` 100 times over.
4. Edit [`Text._get_layout`](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/text.py#L329) to print out whether the cache hit worked or failed.

In the end, you will get about 2 or 3 failures out of the 500. On the failures, the cache returns a value. It turns out that every once in a while, the renderer will have the same ID (or be the same one, I don't know). Since the cache key does not include the mathtext fontset, the cached value will be for some other one and thus be in the wrong place.

I'm not sure how to easily fix this. Ideally, the fontset would be part of the key, but it's not actually used by `Text` objects, but the `MathTextParser`. According to #7107, baking the `rcParam` directly into initialization is a hard problem. Maybe we should find a way to disable the cache during test?

### PR DESCRIPTION
**Pull request by QuLogic**
As noted in #7911, the mathtext tests sometimes fail if the `id` of the renderer is the same between tests. Since it's difficult to add the mathtext fontset to the key, I fixed/hacked it into working by adding a cache-busting "unique" value to each renderer, which is just a globally-incremented number. I thought of using a UUID or random number, but that seemed too expensive since we don't need to worry about threading locks on globals (AFAIK).

I ran the test from #7911, but using the repeat plugin to run everything 100 times over (so 50000 tests total.) So I'm not 100% certain, but I believe this fixes #7911.

Additionally, we've lately been seeing several failures of `test_savefig_to_stringio` on AppVeyor. This doesn't usually fail locally, but seems to be a timeout on the TeX cache. I've added some retries on these tests to hopefully reduce the problem.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/whats_new.rst if major new feature
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by QuLogic** [comment]
Huh, just realized that @tacaswell outlined basically this procedure in #7911 but I didn't recall it; must have been stuck in my subconscious somewhere.

**Comment by tacaswell** [comment]
I don't think this is thread safe, and that is ok.  I do not think we make the claim that Matplotlib is threadsafe in any way.  Further most of our time is spent in python and I do not think we release the GIL in  the c extensions so it is not clear to me you would get any major gains (unless you really are I/O bound).

Although the way that this will fail to be thread safe also means both objects existing at the same time in which case the `id()` value should be different.  If the race condition is between two threads calling `_uid` on the same renderer at the same time then the user is trying to render the same object on more than one thread and I expect other things to break.

<details>
There are two places to get race conditions here (see 
https://www.youtube.com/watch?v=7SSYhuk5hmc for lots more details)  If you look 

```python
x = 5


class Test:
    
    def foo():
        global x
        x += 1
        self.cache = x

```
and

```python
In [91]: t = Test()

In [93]: dis.dis(t.foo)
  8           0 LOAD_GLOBAL              0 (x)
              2 LOAD_CONST               1 (1)
              4 INPLACE_ADD
              6 STORE_GLOBAL             0 (x)

  9           8 LOAD_GLOBAL              0 (x)
             10 LOAD_GLOBAL              1 (self)
             12 STORE_ATTR               2 (cache)
             14 LOAD_CONST               0 (None)
             16 RETURN_VALUE
```

I think you can get races conditions between all of the op-codes in line 8 which missing an increment and between lines 8 and 9 (both increment, but then both use the higher value).
<\details>

**Comment by anntzer** [comment]
Should this just go into `RendererBase.__hash__`? (then `get_prop_tup` would just stick `renderer` into the tuple instead of `id(renderer), renderer._uid`.)

**Comment by tacaswell** [comment]
We probably do not want to go down the `__hash__` route without implementing all of other `__eq__` and friends methods.

**Comment by QuLogic** [comment]
For thread-safety, I was thinking more of subsequent renderers. Across the threads, the renderers would have different `id` anyway, so it'd be fine if they ended up with the same value.

Theoretically, thread 1 could create a couple renderers in the time it takes thread 2 to store the wrong value. Then the next time thread 1 created a new `id`-conflicting renderer it would have the wrong `uid`. But something would have to be seriously wrong with thread 2 if thread 1 was able to create multiple renderers in the time it took to load+add+store.

**Comment by anntzer** [comment]
Actually you can just use itertools.count as a threadsafe counter (create _counter at the module level and call `next(_counter)` to get a uid).  Yes, this is only due to a CPython implementation detail (i.e. the GIL) but it's not worse than the current PR and should be at least as fast (or better) (in any case I doubt that's the slowest part of rendering).

**Comment by QuLogic** [comment]
That's a good idea; even if not thread-safe on all implementation, it's at least clearer.

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -212,10 +219,23 @@ class RendererBase(object):
...
+    @property
+    def _uid(self):
+        """
+        A lightweight id for unique-ification purposes.
+
+        Along with id(self), this combination should be unique enough to use as
+        part of a caching key.
+        """
+        if self._id is None:
+            self._id = next(_unique_renderer_id)
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: I would just make this a regular attribute created at instantiation (and leave a comment in the ctor).  I'm not sure what making this a property buys you...

Comment (QuLogic) [inline lib/matplotlib/backend_bases.py]: I meant to remove this now when switching to itertools but forgot about it.

**Comment by QuLogic** [comment]
@tomspur This patch should help with https://bugzilla.redhat.com/show_bug.cgi?id=1401267 (though leave out the pytest parts if on 1.5.3.)

**Comment by anntzer** [comment]
thanks!

**Comment by matthew-brett** [comment]
@QuLogic - great - thanks for doing that.

**Pull request merged**